lib/citeproc/ruby/renderer/date.rb

```diff
@@ -30,56 +28,91 @@ def render_date(item, node)
           parts, delimiter = node.parts, node.delimiter
         end
 
-        parts.map { |part|
-          render date, part
+        if date.range?
+          render_date_range date, node, parts, delimiter
+        else
+          parts.map { |part|
+            render date, part
+          }.reject(&:empty?).join(delimiter)
+        end
+      end
+
+      def render_date_range(date, node, parts, delimiter)
+        delimit_range_at = 'day' # TODO date.delimit_range_at
+
+        from_parts = parts.reverse.drop_while { |part|
+          part.name != delimit_range_at
+        }.reverse
+
+        unless from_parts.empty?
+          from = from_parts.map { |part|
+            format! render_date_part(date, part, part: 0), part
+          }.reject(&:empty?).join(delimiter)
+
+          suffix, range_delimiter = from_parts[-1].values_at(:suffix, :'range-delimiter')
+
+          unless suffix.nil?
+            from = from[0, from.length - suffix.length]
+          end
+        end
+
+        to = parts.map { |part|
+          format! render_date_part(date, part, part: 1), part
         }.reject(&:empty?).join(delimiter)
+
+        [from, to].join(range_delimiter || '–')
       end
 
+
       # @param date [CiteProc::Date]
       # @param node [CSL::Style::DatePart, CSL::Locale::DatePart]
       # @return [String]
-      def render_date_part(date, node)
+      def render_date_part(date, node, part: 0)
```

Comment:
The signature of this should be changed to take a `CiteProc::DatePart` for `date` once season ranges have been addressed (see below).